File: python/d3gate/jev.py
"""Minimal Jev HTTP client.

Mirrors the opencode jev-tool plugin contract: Bearer auth, a single POST to
/v1/systemone with {model, state, questions}, and retries on 429/529 with
exponential backoff. Batching rule: ALL questions go in ONE request.
"""

from __future__ import annotations

import json
import os
import random
import time
from pathlib import Path
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

JEV_BASE_URL = "https://api.typesafe.ai/v1/systemone"
DEFAULT_MODEL = "jev-latest"
KEY_RETRIES = 4
RETRY_STATUSES = {429, 529}

KEY_FILE = Path.home() / ".config/opencode/.secrets/jev.key"

LEDGER_DEFAULT = Path.home() / ".local/share/opencode/jev-usage.jsonl"


def _append_ledger(body: dict[str, Any], model: str, state: str, questions: dict[str, Any]) -> None:
    """Best-effort append-only usage record; never raises."""
    try:
        ledger = os.environ.get("D3GATE_JEV_LEDGER") or str(LEDGER_DEFAULT)
        usage = body.get("usage") or {}
        record = {
            "time": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "model": model,
            "questions": len(questions),
            "state_chars": len(state),
            "input_tokens": usage.get("input_tokens"),
            "output_tokens": usage.get("output_tokens"),
            "total_tokens": usage.get("total_tokens"),
        }
        with open(ledger, "a", encoding="utf-8") as handle:
            handle.write(json.dumps(record) + "\n")
    except Exception:
        pass


class JevError(RuntimeError):
    pass


def api_key() -> str:
    env = os.environ.get("JEV_API_KEY")
    if env and env.strip():
        return env.strip()
    if KEY_FILE.exists():
        key = KEY_FILE.read_text(encoding="utf-8").strip()
        if key:
            return key
    raise JevError(
        "JEV_API_KEY not set and ~/.config/opencode/.secrets/jev.key is missing"
    )


def evaluate(
    state: str,
    questions: dict[str, Any],
    model: str = DEFAULT_MODEL,
    retries: int = KEY_RETRIES,
) -> dict[str, Any]:
    """POST the full batch of typed questions over one state."""
    payload = json.dumps({"model": model, "state": state, "questions": questions}).encode()
    last_error: JevError | None = None
    for attempt in range(retries):
        if attempt:
            seconds = 0.5 * 2 ** (attempt - 1) + random.uniform(0, 0.25)
            time.sleep(seconds)
        request = Request(
            JEV_BASE_URL,
            data=payload,
            method="POST",
            headers={"Authorization": f"Bearer {api_key()}", "Content-Type": "application/json"},
        )
        try:
            with urlopen(request, timeout=60) as response:
                body = json.load(response)
                if not isinstance(body, dict) or "answers" not in body:
                    raise JevError(f"Unexpected Jev response shape: {json.dumps(body)[:300]}")
                _append_ledger(body, model, state, questions)
                return body
        except HTTPError as error:
            if error.code in RETRY_STATUSES:
                last_error = JevError(f"Jev API {error.code}: {error.read()[:300]}")
                continue
            raise JevError(f"Jev API {error.code}: {error.read()[:300]}")
        except (URLError, TimeoutError) as error:
            last_error = JevError(f"Jev API request failed: {error}")
    raise JevError(f"Jev API retries exhausted: {last_error}")


class JevClient:
    def __init__(self, model: str = DEFAULT_MODEL) -> None:
        self.model = model

    def evaluate(self, state: str, questions: dict[str, Any]) -> dict[str, Any]:
        return evaluate(state, questions, model=self.model)

File: tools/teamflow/jev.mjs
/** Batched Jev evaluation client (JS port of python/d3gate/jev.py). */
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { pathToFileURL } from 'node:url';

const BASE = 'https://api.typesafe.ai/v1/systemone';
const RETRY_STATUSES = new Set([429, 529]);
const KEY_FILE = path.join(os.homedir(), '.config/opencode/.secrets/jev.key');
const LEDGER_DEFAULT = path.join(os.homedir(), '.local/share/opencode/jev-usage.jsonl');

export function jevKey() {
  const env = process.env.JEV_API_KEY?.trim();
  if (env) return env;
  if (fs.existsSync(KEY_FILE)) {
    const k = fs.readFileSync(KEY_FILE, 'utf8').trim();
    if (k) return k;
  }
  throw new Error('JEV_API_KEY not set and ~/.config/opencode/.secrets/jev.key is missing');
}

function appendLedger(body, model, state, questions) {
  try {
    const ledger = process.env.D3GATE_JEV_LEDGER || LEDGER_DEFAULT;
    const usage = body?.usage || {};
    fs.mkdirSync(path.dirname(ledger), { recursive: true });
    fs.appendFileSync(ledger, JSON.stringify({
      time: new Date().toISOString(),
      model,
      questions: Object.keys(questions).length,
      state_chars: state.length,
      input_tokens: usage.input_tokens ?? null,
      output_tokens: usage.output_tokens ?? null,
      total_tokens: usage.total_tokens ?? null,
    }) + '\n');
  } catch { /* best-effort */ }
}

export async function evaluate(state, questions, { model = 'jev-latest', retries = 4 } = {}) {
  const payload = JSON.stringify({ model, state, questions });
  let last;
  for (let attempt = 0; attempt < retries; attempt++) {
    if (attempt) {
      const seconds = 0.5 * 2 ** (attempt - 1) + Math.random() * 0.25;
      await new Promise(r => setTimeout(r, seconds * 1000));
    }
    try {
      const res = await fetch(BASE, {
        method: 'POST',
        headers: { Authorization: `Bearer ${jevKey()}`, 'Content-Type': 'application/json' },
        body: payload,
      });
      if (RETRY_STATUSES.has(res.status)) {
        last = new Error(`Jev API ${res.status}: ${(await res.text()).slice(0, 300)}`);
        continue;
      }
      if (!res.ok) throw new Error(`Jev API ${res.status}: ${(await res.text()).slice(0, 300)}`);
      const body = await res.json();
      if (!(body && typeof body === 'object' && 'answers' in body)) {
        throw new Error(`Unexpected Jev response shape: ${JSON.stringify(body).slice(0, 300)}`);
      }
      appendLedger(body, model, state, questions);
      return body;
    } catch (err) {
      if (!(err && err.message && err.message.startsWith('Jev API'))) throw err;
      last = err;
    }
  }
  throw last || new Error('Jev retries exhausted');
}

export function scored(answers, id) {
  return answers?.[id] ?? null;
}

export function numeric(answer) {
  if (!answer) return null;
  if (typeof answer.noul === 'number') return answer.noul;
  if (typeof answer.probability === 'number') return answer.probability;
  if (typeof answer.value === 'number') return answer.value;
  if (typeof answer.score === 'number') return answer.score;
  if (answer.score && typeof answer.score === 'object' && typeof answer.score.value === 'number') return answer.score.value;
  // Live shape: { choice: "opt", confidence, probabilities: { opt: p } }.
  if (typeof answer.choice === 'string') {
    return typeof answer.confidence === 'number' ? answer.confidence : answer.probabilities?.[answer.choice] ?? null;
  }
  if (answer.choice && typeof answer.choice === 'object') {
    return Math.max(0, ...Object.values(answer.choice).filter(v => typeof v === 'number'));
  }
  return null;
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  const cmd = process.argv[2];
  if (cmd !== 'evaluate') {
    console.error('Use: node tools/teamflow/jev.mjs evaluate');
    process.exit(1);
  }
  let raw = '';
  process.stdin.setEncoding('utf8');
  process.stdin.on('data', (d) => { raw += d; });
  process.stdin.on('end', async () => {
    try {
      const { state, questions, model } = JSON.parse(raw);
      const body = await evaluate(state, questions, { model: model || 'jev-latest' });
      console.log(JSON.stringify(body));
    } catch (err) {
      console.error(err.message || String(err));
      process.exit(1);
    }
  });
}